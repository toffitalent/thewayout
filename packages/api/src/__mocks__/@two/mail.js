const send = jest.fn().mockResolvedValue(undefined);

module.exports = {
  send,
};
