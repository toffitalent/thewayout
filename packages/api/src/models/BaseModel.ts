import { PaginateQueryBuilder, Serialize } from '@disruptive-labs/objection-plugins';
import { Model, ModelOptions, QueryContext } from 'objection';
import { knex } from '../db';

// Bind knex instance to objection
Model.knex(knex);

export class BaseModel extends Serialize(Model) {
  static QueryBuilder = PaginateQueryBuilder;
  QueryBuilderType!: PaginateQueryBuilder<this>;

  static useLimitInFirst = true;

  createdAt!: Date;
  updatedAt!: Date;

  async $beforeInsert(queryContext: QueryContext) {
    await super.$beforeInsert(queryContext);

    this.createdAt = new Date();
    this.updatedAt = this.createdAt;
  }

  async $beforeUpdate(opt: ModelOptions, queryContext: QueryContext) {
    await super.$beforeUpdate(opt, queryContext);

    this.updatedAt = new Date();
  }
}
