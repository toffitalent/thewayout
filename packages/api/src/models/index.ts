export * from './App';
export * from './AppClient';
export * from './AppToken';
export * from './AppUser';
export * from './BaseModel';
export * from './Client';
export * from './Employer';
export * from './Job';
export * from './JobApplication';
export * from './Rsp';
export * from './RspAccount';
export * from './RspInvitation';
export * from './serializers';
export * from './User';
export {
  CheckViolationError,
  ConstraintViolationError,
  DataError,
  DBError,
  ForeignKeyViolationError,
  NotFoundError,
  NotNullViolationError,
  raw,
  ref,
  UniqueViolationError,
  ValidationError,
} from 'objection';
