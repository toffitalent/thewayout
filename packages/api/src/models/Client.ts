import {
  Age,
  Education,
  Ethnicity,
  ExperienceSkills,
  Gender,
  JusticeStatus,
  Languages,
  License,
  MaritalStatus,
  OffenseCategory,
  Orientation,
  PersonalStrengths,
  ReferredBy,
  RelativeExperience,
  ReleasedAt,
  Religion,
  State,
  StateFederal,
  Support,
  TimeServed,
  VeteranBranchOfService,
  VeteranCampaign,
  VeteranOrJustice,
  VeteranRank,
  VeteranStatus,
  VeteranTypeDischarge,
} from '@two/shared';
import { BaseModel } from './BaseModel';
import { ClientSerializer } from './serializers';

export class Client extends BaseModel {
  static tableName = 'clients';
  static Serializer = ClientSerializer;

  readonly id!: string;

  userId!: string;
  firstName!: string;
  lastName!: string;
  justiceStatus!: JusticeStatus;
  facility!: string | null;
  expectedReleasedAt!: string | null;
  releasedCounty!: string | null;
  phone!: string | null;
  address!: string;
  city!: string;
  state!: State | string;
  postalCode!: string;
  support!: Support[];
  gender!: Gender | null;
  orientation!: Orientation | null;
  religion!: Religion | null;
  maritalStatus!: MaritalStatus | null;
  age!: Age | null;
  disability!: boolean | null;
  ethnicity!: Ethnicity | null;
  veteranStatus!: VeteranStatus | null;
  referredBy!: ReferredBy;
  personalStrengths!: PersonalStrengths[];
  experience!: ExperienceSkills[];
  languages!: Languages[];
  offense!: OffenseCategory[] | null;
  sexualOffenderRegistry!: boolean | null;
  sbn!: boolean | null;
  timeServed!: TimeServed | null;
  releasedAt!: ReleasedAt | null;
  stateOrFederal!: StateFederal | null;
  relativeExperience!: RelativeExperience[];
  personalSummary!: string;
  education!: Education[];
  license!: License[];
  veteranOrJustice!: VeteranOrJustice[];
  veteranService?: VeteranBranchOfService[];
  veteranRank?: VeteranRank;
  veteranReservist?: boolean;
  veteranCampaigns?: VeteranCampaign[];
  veteranStartAt?: string;
  veteranEndAt?: string;
  veteranTypeDischarge?: VeteranTypeDischarge;
  veteranDd214?: boolean;
}
