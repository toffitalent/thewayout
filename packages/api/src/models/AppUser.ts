import type { RelationMappingsThunk } from 'objection';
import { App } from './App';
import { BaseModel } from './BaseModel';
import { AppUserSerializer } from './serializers';
import { User } from './User';

export class AppUser extends BaseModel {
  static tableName = 'app_users';
  static Serializer = AppUserSerializer;

  readonly id!: string;
  appId!: string;
  userId!: string;
  scope!: string[];

  app?: App;
  user?: User;

  static relationMappings: RelationMappingsThunk = () => ({
    app: {
      relation: BaseModel.BelongsToOneRelation,
      modelClass: App,
      join: {
        from: 'app_users.appId',
        to: 'apps.id',
      },
    },

    user: {
      relation: BaseModel.BelongsToOneRelation,
      modelClass: User,
      join: {
        from: 'app_users.userId',
        to: 'users.id',
      },
    },
  });
}
