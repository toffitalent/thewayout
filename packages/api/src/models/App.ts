import type { RelationMappingsThunk } from 'objection';
import { AppClient } from './AppClient';
import { AppToken } from './AppToken';
import { AppUser } from './AppUser';
import { BaseModel } from './BaseModel';
import { AppSerializer } from './serializers';

export class App extends BaseModel {
  static tableName = 'apps';
  static Serializer = AppSerializer;

  readonly id!: string;
  name!: string;
  userId!: string | null;

  clients?: AppClient[];
  tokens?: AppToken[];
  users?: AppUser[];

  static relationMappings: RelationMappingsThunk = () => ({
    clients: {
      relation: BaseModel.HasManyRelation,
      modelClass: AppClient,
      join: {
        from: 'apps.id',
        to: 'app_clients.appId',
      },
    },

    tokens: {
      relation: BaseModel.HasManyRelation,
      modelClass: AppToken,
      join: {
        from: 'apps.id',
        to: 'app_tokens.appId',
      },
    },

    users: {
      relation: BaseModel.HasManyRelation,
      modelClass: AppUser,
      join: {
        from: 'apps.id',
        to: 'app_users.appId',
      },
    },
  });
}
