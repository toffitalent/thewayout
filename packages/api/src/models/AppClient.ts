import type { RelationMappingsThunk } from 'objection';
import { AuthGrantType } from '@two/shared';
import { App } from './App';
import { AppToken } from './AppToken';
import { BaseModel } from './BaseModel';

export class AppClient extends BaseModel {
  static tableName = 'app_clients';

  readonly id!: string;
  appId!: string;
  name!: string;
  secret!: string;
  grantType!: AuthGrantType;
  redirectUri!: string | null;

  app?: App;
  tokens?: AppToken[];

  static relationMappings: RelationMappingsThunk = () => ({
    app: {
      relation: BaseModel.BelongsToOneRelation,
      modelClass: App,
      join: {
        from: 'app_clients.appId',
        to: 'apps.id',
      },
    },

    tokens: {
      relation: BaseModel.HasManyRelation,
      modelClass: AppToken,
      join: {
        from: 'app_clients.id',
        to: 'app_tokens.clientId',
      },
    },
  });
}
