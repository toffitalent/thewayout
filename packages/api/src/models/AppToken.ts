import type { RelationMappingsThunk } from 'objection';
import { App } from './App';
import { AppClient } from './AppClient';
import { BaseModel } from './BaseModel';
import { User } from './User';

export class AppToken extends BaseModel {
  static tableName = 'app_tokens';

  readonly id!: string;
  appId!: string;
  clientId!: string;
  userId!: string;
  token!: Buffer;
  self!: boolean;

  app?: App;
  client?: AppClient;
  user?: User;

  static relationMappings: RelationMappingsThunk = () => ({
    app: {
      relation: BaseModel.BelongsToOneRelation,
      modelClass: App,
      join: {
        from: 'app_tokens.appId',
        to: 'apps.id',
      },
    },

    client: {
      relation: BaseModel.BelongsToOneRelation,
      modelClass: AppClient,
      join: {
        from: 'app_tokens.clientId',
        to: 'app_clients.id',
      },
    },

    user: {
      relation: BaseModel.BelongsToOneRelation,
      modelClass: User,
      join: {
        from: 'app_tokens.userId',
        to: 'users.id',
      },
    },
  });
}
