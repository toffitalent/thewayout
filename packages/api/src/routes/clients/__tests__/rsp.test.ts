import { faker } from '@faker-js/faker';
import { OffenseCategory, Support } from '@two/shared';
import { Rsp } from '@app/models';
import { auth, fixtures, request, resetDb } from '@test';

const rsps = [
  {
    name: '1',
    offenses: [OffenseCategory.arson],
    support: [Support.basicTechSkills],
    servicesArea: ['Adams'],
  },
  {
    name: '2',
    offenses: [OffenseCategory.bailJumping],
    support: [Support.childCare],
    servicesArea: ['Douglas'],
  },
  {
    name: '3',
    offenses: [OffenseCategory.diuDwi],
    support: [Support.employment],
    servicesArea: ['Pepin'],
  },
];

describe('Routes > Clients > Rsp', () => {
  beforeEach(async () => {
    await resetDb();

    faker.seed(1);
  });

  describe('GET /api/v1/clients/:id/rsp', () => {
    beforeEach(async () => {
      await Rsp.query().insert(rsps);
    });

    test('list rsp for clients', async () => {
      const res = await request()
        .get(`/api/v1/clients/${fixtures.clientProfile1.id}/rsp`)
        .use(auth(fixtures.user4.id, ['client']))
        .query({
          support: [Support.clothing],
          offenses: [OffenseCategory.burglary],
          county: 'Adams',
        });

      expect(res.status).toBe(200);
      expect(res.type).toBe('application/json');
      expect(res.body).toMatchSnapshot();
    });

    test('lists all rsp when no filters added', async () => {
      const res = await request()
        .get(`/api/v1/clients/${fixtures.clientProfile1.id}/rsp`)
        .use(auth(fixtures.user4.id, ['client']));

      expect(res.status).toBe(200);
      expect(res.body.length).toBe(5);
    });

    test('lists only rsp that match client offenses', async () => {
      const res1 = await request()
        .get(`/api/v1/clients/${fixtures.clientProfile1.id}/rsp`)
        .use(auth(fixtures.user4.id, ['client']))
        .query({
          offenses: [OffenseCategory.arson],
        });

      expect(res1.status).toBe(200);
      expect(res1.body.length).toBe(1);
      expect(res1.body[0].name).toBe('1');
      expect({ body: res1.body }).toMatchSnapshot({ body: [{ id: expect.any(String) }] });

      const res2 = await request()
        .get(`/api/v1/clients/${fixtures.clientProfile1.id}/rsp`)
        .use(auth(fixtures.user4.id, ['client']))
        .query({
          offenses: [OffenseCategory.arson, OffenseCategory.embezzlement],
        });

      expect(res2.status).toBe(200);
      expect(res2.body.length).toBe(0);
      expect(res2.body).toMatchSnapshot();
    });

    test('lists only rsp that match client county', async () => {
      const res = await request()
        .get(`/api/v1/clients/${fixtures.clientProfile1.id}/rsp`)
        .use(auth(fixtures.user4.id, ['client']))
        .query({
          county: 'Douglas',
        });

      expect(res.status).toBe(200);
      expect(res.body.length).toBe(1);
      expect(res.body[0].name).toBe('2');
      expect({ body: res.body }).toMatchSnapshot({ body: [{ id: expect.any(String) }] });
    });

    test('lists only rsp that match client support', async () => {
      const res = await request()
        .get(`/api/v1/clients/${fixtures.clientProfile1.id}/rsp`)
        .use(auth(fixtures.user4.id, ['client']))
        .query({
          support: [Support.employment, Support.domesticViolence],
        });

      expect(res.status).toBe(200);
      expect(res.body.length).toBe(1);
      expect(res.body[0].name).toBe('3');
      expect({ body: res.body }).toMatchSnapshot({ body: [{ id: expect.any(String) }] });
    });
  });
});
