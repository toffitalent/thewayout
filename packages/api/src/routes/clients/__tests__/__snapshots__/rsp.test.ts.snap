// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Routes > Clients > Rsp GET /api/v1/clients/:id/rsp list rsp for clients 1`] = `
[
  {
    "description": "repellendus maiores sunt",
    "id": "01817e24-0d43-7ab4-927f-49d0b06d5cf2",
    "name": "Funk, Koch and Wiegand",
    "servicesArea": [
      "Adams",
      "Ashland",
      "Barron",
    ],
    "support": [
      "basicHygieneSupport",
      "clothing",
      "basicTechSkills",
      "childCare",
    ],
  },
  {
    "description": "Enim sequi iusto.
Explicabo quisquam omnis nisi corporis corrupti.
Cupiditate qui culpa numquam iure quisquam vero architecto.",
    "id": "01817e24-0d42-77e5-9a94-ad514d8ec5dc",
    "name": "Oberbrunner and Sons",
    "servicesArea": [
      "Adams",
      "Ashland",
      "Barron",
    ],
    "support": [
      "basicHygieneSupport",
      "clothing",
      "basicTechSkills",
      "childCare",
    ],
  },
]
`;

exports[`Routes > Clients > Rsp GET /api/v1/clients/:id/rsp lists only rsp that match client county 1`] = `
{
  "body": [
    {
      "description": null,
      "id": Any<String>,
      "name": "2",
      "servicesArea": [
        "Douglas",
      ],
      "support": [
        "childCare",
      ],
    },
  ],
}
`;

exports[`Routes > Clients > Rsp GET /api/v1/clients/:id/rsp lists only rsp that match client offenses 1`] = `
{
  "body": [
    {
      "description": null,
      "id": Any<String>,
      "name": "1",
      "servicesArea": [
        "Adams",
      ],
      "support": [
        "basicTechSkills",
      ],
    },
  ],
}
`;

exports[`Routes > Clients > Rsp GET /api/v1/clients/:id/rsp lists only rsp that match client offenses 2`] = `[]`;

exports[`Routes > Clients > Rsp GET /api/v1/clients/:id/rsp lists only rsp that match client support 1`] = `
{
  "body": [
    {
      "description": null,
      "id": Any<String>,
      "name": "3",
      "servicesArea": [
        "Pepin",
      ],
      "support": [
        "employment",
      ],
    },
  ],
}
`;
